### ISSUE TITLE
compiler: remove unused struct fields

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#21010]
Unfortunately we cannot have a compile error for unused struct fields (#12854) since whether a field is used depends on semantic analysis (#3028). However, for structs that have automatic layout, it is legal for the compiler to remove fields that are not used. The compiler should do that automatically.

**Comment by silversquirl** [comment]
Wouldn't this mean no function can be sent to code generation until all functions have been analyzed? Since a function analyzed later could use the field and thus affect the codegen of previous functions.
Seems like the compile time performance implications would be pretty significant

**Comment by andrewrk** [comment]
That's a great point, and since it might have compiler design implications, it means it needs to be tackled earlier.

**Comment by mlugg** [comment]
Yeah, this isn't feasible as-is for precisely the reason silversquirl gave above. Moreover, any semantic analysis which requires resolving the type layout -- for instance, `@sizeOf` -- would require first resolving all possible usages of the type. So, we have an absurd situation where `@sizeOf` logically depends on the entire program -- including the code following itself -- being analyzed.

**Comment by mlugg** [comment]
If we go ahead with Martin's idea to make `@field` and friends take an `@Type(.EnumLiteral)` rather than a string, then we could define a *sufficient* condition for omitting a field: if that identifier never occurs in any other type field or enum literal. However, this condition is so narrow as to be entirely useless.

**Comment by andrewrk** [comment]
It would be possible to combine these two things:
* compile error for unused field (based on semantic analysis results)
* ability to declare any field unused (including in transitive dependencies)

in order to trade programmer effort for optimality. It would probably be a huge pain in the ass though.

**Comment by mlugg** [comment]
Hmm -- so for every field, you can mark it (fictional syntax) `unused(is_field_unused)` (default is `unused(false)`), and you get a compile error if it was used and marked unused or vice versa. Yeah, that could work, but I anticipate it being a huge PITA. And it makes the benefits of this proposal dubious -- at that point, just make the field type conditional, marking it `void` if it'll be unused.

**Comment by andrewrk** [comment]
My point is that the unused declaration would exist potentially even outside the module that declares the struct. The struct would be declared like normal (consider the motivating use case of #20977).
